### PR TITLE
[Feature] Completely resolving the issue of query failures in Hive external tables due to inconsistent metadata caching

### PR DESCRIPTION
**Pull request by stephen-shelby**
Fixes #issue
If modifications are made to the partition files of a Hive table by user, such as through "insert overwrite partition", the Frontend couldn't be aware of these changes. As a result, queries may use the old file information cached in the FE for execution. Be will throw exception when accessing a non-exist remote files. So we need to resolve this issue.
There are some alternative solutions:
- aws sqs/sns  // Cannot be aware of a complete transaction.
- glue notification // it can't work when user use hive metastore.
- hive metastore listener // user need to enable metastore events in the hive metastore config and restart it. most of user can't do this operation.
- Background Periodic Refresh  // The refresh interval is long, and users still have a high probability of encoutering query errors. The refresh interval is short, background tasks put a lot of pressure on hive metastore and namenode.

I think the most out-of-the-box solution is to retry the query. When the file can't be found, refresh the metadata cache and regenerate the scan range.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 46 (32.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|22|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 492, 494, 497, 499, 501, 502, 504, 506, 507, 508, 509]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

File: be/src/fs/hdfs/fs_hdfs.cpp [hunk 1]
```diff
@@ -461,7 +461,7 @@ StatusOr<std::unique_ptr<WritableFile>> HdfsFileSystem::new_writable_file(const
 
     hdfsFile file = hdfsOpenFile(hdfs_client->hdfs_fs, path.c_str(), flags, hdfs_write_buffer_size, 0, 0);
     if (file == nullptr) {
-        return Status::InternalError(fmt::format("hdfsOpenFile failed, file={}", path));
```
Comment (Smith-Cruise) [inline be/src/fs/hdfs/fs_hdfs.cpp:464]: What if user configure wrong credentials, will it be retried also?
Can you just handle 404 situation?

Comment (Smith-Cruise) [reply be/src/fs/hdfs/fs_hdfs.cpp:464]: https://github.com/apache/hadoop/blob/9acf462d268a7dee8d031978add0034fcf77f861/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/exception.c#L41

Comment (Smith-Cruise) [reply be/src/fs/hdfs/fs_hdfs.cpp:464]: https://github.com/apache/hadoop/blob/9acf462d268a7dee8d031978add0034fcf77f861/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/hdfs.c#L4263

Comment (Smith-Cruise) [reply be/src/fs/hdfs/fs_hdfs.cpp:464]: just some references.

Comment (stephen-shelby) [reply be/src/fs/hdfs/fs_hdfs.cpp:464]: good catch

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 46 (32.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|22|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 492, 494, 497, 499, 501, 502, 504, 506, 507, 508, 509]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

File: be/src/connector/hive_connector.cpp [hunk 2]
```diff
@@ -488,6 +488,9 @@ Status HiveDataSource::_init_scanner(RuntimeState* state) {
     Status st = scanner->open(state);
     if (!st.ok()) {
         auto msg = fmt::format("file = {}", native_file_path);
+        if (st.is_io_error() && st.message().starts_with("code=404")) {
```
Comment (Smith-Cruise) [inline be/src/connector/hive_connector.cpp:491]: Add some comments, we catch the file not found error from aws s3 sdk.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 46 (32.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|22|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 492, 494, 497, 499, 501, 502, 504, 506, 507, 508, 509]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 47 (31.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|23|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 492, 494, 497, 499, 501, 502, 504, 506, 507, 508, 509, 511]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 51 (29.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|27|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 494, 496, 499, 501, 502, 505, 507, 508, 510, 512, 513, 514, 515, 517]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 3]
```diff
@@ -464,6 +469,52 @@ public void execute() throws Exception {
 
                         handleQueryStmt(execPlan);
                         break;
+                    } catch (StarRocksConnectorException e) {
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:472]: Is it possible that some `StarRocksConnectorException` which isn't caused by expired meta cache arriving here? Which may cause the query is also retried when some non-recovable error occurs.

Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java]: it's impossible currently. But I have defined another type exception to catch it.

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 15 / 59 (25.42%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|5|00.00%|[40, 43, 71, 72, 88]|
|:large_blue_circle:|com/starrocks/common/Status.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|2|00.00%|[1393, 1394]|
|:large_blue_circle:|com/starrocks/connector/exception/RemoteFileNotFoundException.java|0|8|00.00%|[21, 22, 25, 26, 29, 30, 33, 38]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|27|00.00%|[472, 479, 480, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 494, 496, 499, 501, 502, 505, 507, 508, 510, 512, 513, 514, 515, 517]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|12|13|92.31%|[53]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|3|3|100.00%|[]|

**Review by Youngwb — APPROVED** [review]

**Comment by stephen-shelby** [comment]
https://github.com/Mergifyio backport branch-3.1

**Comment by stephen-shelby** [comment]
https://github.com/Mergifyio backport branch-3.0

**Pull request merged**